python/src/iceberg/expression/literals.py

```diff
@@ -0,0 +1,654 @@
+
+
+@literal.register
+def _(value: str):
+    return StringLiteral(value)
+
+
+@literal.register
+def _(value: uuid.UUID):
+    return UUIDLiteral(value)
+
+
+@literal.register
+def _(value: bytes):
+    return FixedLiteral(value)
+
+
+@literal.register
+def _(value: bytearray):
+    return BinaryLiteral(value)
+
+
+@literal.register
+def _(value: Decimal):
+    return DecimalLiteral(value)
+
+
+class ComparableLiteral(BaseLiteral):
+    def __init__(self, repr_string: str, value):
+        super(ComparableLiteral, self).__init__(repr_string, value)
+
+    def to(self, type_var):
+        raise NotImplementedError()
+
+    def __eq__(self, other):
+        return self.value == other.value
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+    def __lt__(self, other):
+        return self.value < other.value
+
+    def __gt__(self, other):
+        return self.value > other.value
+
+    def __le__(self, other):
+        return self.value <= other.value
+
+    def __ge__(self, other):
+        return self.value >= other.value
+
+
+class AboveMax(Literal, Singleton):
+    def value(self):
+        raise RuntimeError("AboveMax has no value")
+
+    def to(self, type_var):
+        raise RuntimeError("Cannot change the type of AboveMax")
+
+    def __str__(self):
+        return "aboveMax"
+
+
+class BelowMin(Literal, Singleton):
+    def value(self):
+        raise RuntimeError("BelowMin has no value")
+
+    def to(self, type_var):
+        raise RuntimeError("Cannot change the type of BelowMin")
+
+    def __str__(self):
+        return "belowMin"
+
+
+class BooleanLiteral(ComparableLiteral):
+    def __init__(self, value):
+        super(BooleanLiteral, self).__init__(f"BooleanLiteral({value})", value)
+
+    @singledispatchmethod
+    def to(self, type_var):
+        return None
+
+    @to.register(BooleanType)
+    def _(self, type_var):
+        return self
+
+
+class IntegerLiteral(ComparableLiteral):
+    def __init__(self, value):
+        super(IntegerLiteral, self).__init__(f"IntegerLiteral({value})", value)
+
+    @singledispatchmethod
+    def to(self, type_var):
+        return None
+
+    @to.register(IntegerType)
+    def _(self, type_var):
+        return self
+
+    @to.register(LongType)
+    def _(self, type_var):
+        return LongLiteral(self.value)
+
+    @to.register(FloatType)
+    def _(self, type_var):
+        return FloatLiteral(self.value)
+
+    @to.register(DoubleType)
+    def _(self, type_var):
+        return DoubleLiteral(self.value)
+
+    @to.register(DateType)
+    def _(self, type_var):
+        return DateLiteral(self.value)
+
+    @to.register(DecimalType)
+    def _(self, type_var):
+        if type_var.scale == 0:
+            return DecimalLiteral(Decimal(self.value))
+        else:
+            return DecimalLiteral(
+                Decimal(self.value).quantize(
+                    Decimal("." + "".join(["0" for _ in range(1, type_var.scale)]) + "1"), rounding=ROUND_HALF_UP
```

Comment:
String manipulation for decimal conversion looks suspicious to me. Is there no API to do this?

Reply:
yeah I was looking into [Decimal](https://docs.python.org/3/library/decimal.html#) and looks like the quantize is the API supposed to round the Decimal value to its first operant. I can kind of see that construct the Decimal from String is for better precision compare to float counterpart. 

I am not very familiar with the Decimal and did not find a better API to do this, I guess we can keep this as of now and come back if someone find a better way to do this ? What do you think

Reply:
perhaps instead of using this string manipulation you can use the init by tuple 
```python
Decimal(self.value).quantize(Decimal((0, (1,), -type_var.scale)), rounding=ROUND_HALF_UP)`
```

Reply:
Thank you @CircArgs , I think using tuple to initialize the Decimal looks much better than the string manipulation. Updated per your suggestion

Reply:
Looks like this one didn't get updated?

Reply:
I think this should be:

```python
sign, digits, _ = Decimal(self.value).to_tuple()
return Decimal((sign, digits, -type_var.scale))
```